packages/@cdktf/cli-core/src/lib/models/deploy-machine.ts

```diff
@@ -81,137 +98,196 @@ export const deployMachine = createMachine<
+                  }),
+                ],
+              },
             },
           },
         },
       },
+      exited: { type: "final" },
+      stopped: { type: "final" },
     },
-    exited: { type: "final" },
-    stopped: { type: "final" },
   },
-});
+  {
+    services: {
+      runTerraformInPty: (context, event) =>
+        terraformPtyService(context, event, spawnPtyFromEvent),
+    },
+  }
+);
+
+function spawnPtyFromEvent(event: DeployEvent): pty.IPty {
+  if (event.type !== "START") {
+    throw Errors.Internal(
+      `Terraform CLI invocation state machine: Unexpected event caused transition to the running state: ${event.type}`
+    );
+  }
+
+  const { args, options } = event.pty;
+  const file =
+    os.platform() === "win32"
+      ? findExecutable(event.pty.file, options.cwd, options)
+      : event.pty.file;
+  logger.trace(
+    `Spawning pty with file=${file}, args=${
+      Array.isArray(args) ? `[${args.join(", ")}]` : `"${args}"`
+    }, options=${JSON.stringify(options)}`
+  );
+  const p = pty.spawn(file, args, options);
+
+  return p;
+}
+
+export function terraformPtyService(
+  _context: DeployContext,
+  event: DeployEvent,
+  spawnFunction: (event: DeployEvent) => pty.IPty
+): (send: Sender<DeployEvent>, onReceive: Receiver<DeployEvent>) => void {
+  return (send: Sender<DeployEvent>, onReceive: Receiver<DeployEvent>) => {
+    const p = spawnFunction(event);
+
+    onReceive((event: DeployEvent) => {
+      if (event.type === "SEND_INPUT") {
+        p.write(event.input);
+      }
+    });
+
+    p.onData((line) => {
+      let hideLine = false;
+
+      // possible events based on line
+      if (line.includes("approved using the UI or API")) {
+        send({ type: "APPROVED_EXTERNALLY" });
+      } else if (line.includes("discarded using the UI or API")) {
+        send({ type: "REJECTED_EXTERNALLY" });
+      } else if (
+        line.includes("Do you want to perform these actions") ||
+        line.includes("Do you really want to destroy all resources?")
+      ) {
+        send({ type: "REQUEST_APPROVAL" });
+      } else if (line.includes("var.") && line.includes("Enter a value:")) {
+        hideLine = true;
+
+        const variableName = extractVariableNameFromPrompt(line);
+        send({
+          type: "LINE_RECEIVED",
+          line: missingVariable(variableName),
+        });
+        send({ type: "VARIABLE_MISSING", variableName });
+      } else if (
+        line.includes("Do you want to override the soft failed policy check?")
+      ) {
+        hideLine = true;
+        send({ type: "LINE_RECEIVED", line });
+
+        send({ type: "REQUEST_SENTINEL_OVERRIDE" });
+      }
+
+      if (!hideLine) {
+        send({
+          type: "LINE_RECEIVED",
+          line,
+        });
+      }
+    });
+    p.onExit(({ exitCode }) => {
+      send({ type: "EXITED", exitCode });
+    });
+
+    return () => {
+      logger.trace("Terracorm CLI state machine: cleaning up pty");
```

Comment:
not your typo (but mine 😁)
```suggestion
      logger.trace("Terraform CLI state machine: cleaning up pty");
```